File: src/ClientLibs/android/PoseidonClient.kt
package com.poseidon.client

/**
 * Kotlin coroutine API for Poseidon pub/sub.
 * All methods are suspend functions that can be called from coroutines.
 */
class PoseidonClient(private val connection: PoseidonConnection) {

    companion object {
        // Method codes matching client_protocol.h
        private const val METHOD_CHANNEL_CREATE = 1
        private const val METHOD_CHANNEL_JOIN = 2
        private const val METHOD_CHANNEL_LEAVE = 3
        private const val METHOD_CHANNEL_DESTROY = 4
        private const val METHOD_CHANNEL_MODIFY = 5
        private const val METHOD_SUBSCRIBE = 6
        private const val METHOD_UNSUBSCRIBE = 7
        private const val METHOD_PUBLISH = 8
        private const val METHOD_ALIAS_REGISTER = 9
        private const val METHOD_ALIAS_UNREGISTER = 10
    }

    suspend fun createChannel(name: String): String {
        val response = connection.sendRequest(METHOD_CHANNEL_CREATE, name)
        return decodeTopicId(response)
    }

    suspend fun joinChannel(topicOrAlias: String): String {
        val response = connection.sendRequest(METHOD_CHANNEL_JOIN, topicOrAlias)
        return decodeTopicId(response)
    }

    suspend fun leaveChannel(topicId: String) {
        connection.sendRequest(METHOD_CHANNEL_LEAVE, topicId)
    }

    suspend fun destroyChannel(topicId: String, ownerKey: ByteArray) {
        connection.sendRequest(METHOD_CHANNEL_DESTROY, topicId, ownerKey)
    }

    suspend fun modifyChannel(topicId: String, config: Map<String, Any>, ownerKey: ByteArray) {
        // Simplified — production would serialize config as CBOR
        connection.sendRequest(METHOD_CHANNEL_MODIFY, topicId, ownerKey)
    }

    suspend fun subscribe(topicPath: String) {
        connection.sendRequest(METHOD_SUBSCRIBE, topicPath)
    }

    suspend fun unsubscribe(topicPath: String) {
        connection.sendRequest(METHOD_UNSUBSCRIBE, topicPath)
    }

    suspend fun publish(topicPath: String, data: ByteArray) {
        connection.sendRequest(METHOD_PUBLISH, topicPath, data)
    }

    suspend fun registerAlias(name: String, topicId: String) {
        connection.sendRequest(METHOD_ALIAS_REGISTER, name, topicId.toByteArray())
    }

    suspend fun unregisterAlias(name: String) {
        connection.sendRequest(METHOD_ALIAS_UNREGISTER, name)
    }

    fun onDelivery(callback: (topicId: String, subtopic: String, data: ByteArray) -> Unit) {
        connection.onDelivery(callback)
    }

    private fun decodeTopicId(response: ByteArray): String {
        // Simplified — production would decode CBOR response
        return String(response, Charsets.UTF_8)
    }
}

File: src/ClientLibs/swift/PoseidonClient.swift
import Foundation

/// Swift async/await API for Poseidon pub/sub.
@available(macOS 10.15, iOS 13.0, *)
class PoseidonClient {
    private let connection: PoseidonConnection

    init(connection: PoseidonConnection) {
        self.connection = connection
    }

    func createChannel(name: String) async throws -> String {
        let response = try await connection.sendRequest(
            method: PoseidonConnection.methodChannelCreate,
            topicPath: name)
        return String(data: response, encoding: .utf8) ?? ""
    }

    func joinChannel(topicOrAlias: String) async throws -> String {
        let response = try await connection.sendRequest(
            method: PoseidonConnection.methodChannelJoin,
            topicPath: topicOrAlias)
        return String(data: response, encoding: .utf8) ?? ""
    }

    func leaveChannel(topicId: String) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodChannelLeave,
            topicPath: topicId)
    }

    func destroyChannel(topicId: String, ownerKey: Data) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodChannelDestroy,
            topicPath: topicId,
            payload: ownerKey)
    }

    func modifyChannel(topicId: String, config: [String: Any], ownerKey: Data) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodChannelModify,
            topicPath: topicId,
            payload: ownerKey)
    }

    func subscribe(topicPath: String) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodSubscribe,
            topicPath: topicPath)
    }

    func unsubscribe(topicPath: String) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodUnsubscribe,
            topicPath: topicPath)
    }

    func publish(topicPath: String, data: Data) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodPublish,
            topicPath: topicPath,
            payload: data)
    }

    func registerAlias(name: String, topicId: String) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodAliasRegister,
            topicPath: name,
            payload: topicId.data(using: .utf8))
    }

    func unregisterAlias(name: String) async throws {
        _ = try await connection.sendRequest(
            method: PoseidonConnection.methodAliasUnregister,
            topicPath: name)
    }

    func onDelivery(_ callback: @escaping (String, String, Data) -> Void) {
        connection.onDelivery(callback)
    }
}